src/libs/SuffixUkkonenTree.ts

```diff
@@ -0,0 +1,310 @@
+ * - This is so that the numbers from a-z are in the range 0-25.
+ * - 26 is for the delimiter character "{",
+ * - 27 is for the end character "|".
+ */
+function stringToArray(input: string) {
+    const res: number[] = [];
+    for (let i = 0; i < input.length; i++) {
+        const charCode = input.charCodeAt(i) - CHAR_CODE_A;
+        if (charCode >= 0 && charCode < ALPHABET_SIZE) {
+            res.push(charCode);
+        }
+    }
+    return res;
+}
+
+const aToZRegex = /[^a-z]/gi;
+// The character that separates the different options in the search string
+const delimiterChar = '{';
+
+type PrepareDataParams<T> = {
+    data: T[];
+    transform: (data: T) => string;
+};
+
+function prepareData<T>({data, transform}: PrepareDataParams<T>): [string, Array<T | undefined>] {
+    const searchIndexList: Array<T | undefined> = [];
+    const str = data
+        .map((option) => {
+            let searchStringForTree = transform(option);
+            // Remove all none a-z chars:
+            searchStringForTree = searchStringForTree.toLowerCase().replace(aToZRegex, '');
+
+            if (searchStringForTree.length > 0) {
+                // We need to push an array that has the same length as the length of the string we insert for this option:
+                const indexes = Array.from({length: searchStringForTree.length}, () => option);
+                // Note: we add undefined for the delimiter character
+                searchIndexList.push(...indexes, undefined);
+            } else {
+                return undefined;
+            }
+
+            return searchStringForTree;
+        })
+        // slightly faster alternative to `.filter(Boolean).join(delimiterChar)`
+        .reduce((acc: string, curr) => {
+            if (!curr) {
+                return acc;
+            }
+
+            if (acc === '') {
+                return curr;
+            }
+
+            return `${acc}${delimiterChar}${curr}`;
+        }, '');
+
+    return [str, searchIndexList];
+}
+
+/**
+ * Makes a tree from an input string
+ * **Important:** As we only support an alphabet of 26 characters, the input string should only contain characters from a-z.
+ * Thus, all input data must be cleaned before being passed to this function.
+ * If you then use this tree for search you should clean your search input as well (so that a search query of "testuser@myEmail.com" becomes "testusermyemailcom").
+ */
+function makeTree<T>(compose: Array<PrepareDataParams<T>>) {
+    const start1 = performance.now();
+    const strings = [];
+    const indexes: Array<Array<T | undefined>> = [];
+
+    for (const {data, transform} of compose) {
+        const [str, searchIndexList] = prepareData({data, transform});
+        strings.push(str);
+        indexes.push(searchIndexList);
+    }
+    const stringToSearch = `${strings.join('')}|`; // End Character
+    console.log('building search strings', performance.now() - start1);
+
+    const a = stringToArray(stringToSearch);
```

Comment:
Just to get ahead of this, we'll definitely want descriptive names for some of these variables